### PR TITLE
[WIP] trying to implement nvdec/vic

### PR DESCRIPTION
**Pull request by namkazt**
Just trying to make it work.
Progress: 
- [ ] nvdec
- [ ] vic

### CODE REVIEW

**Comment by lioncash** [comment]
Our variable naming style uses `snake_case` not `camelCase`

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 1]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+            syncPtValues[syncIncr.syncpt_id] += syncIncr.syncpt_incrs;
+        }
+    }
+
+    LOG_WARNING(
+        Service_NVDRV,
+        "(STUBBED) called, num_cmdbufs: {}, num_relocs: {}, num_syncpt_incrs: {}, num_fences: {}",
+        params.num_cmdbufs, params.num_relocs, params.num_syncpt_incrs, params.num_fences);
+
+    std::memcpy(output.data(), &params, 16);
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
    std::memcpy(output.data(), &params, sizeof(params));
```

**Comment by FernandoS27** [comment]
Btw, yuzu implements sync points if you look at nv_host_ctrl you'll see how to read them.

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 2]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+                sizeof(IoctlRelocShift) * params.num_relocs);
+    startPt += sizeof(IoctlRelocShift) * params.num_relocs;
+    // TODO(namkazt): what to do with this?
+
+    std::vector<IoctlSyncPtIncr> syncPtIncrs(params.num_syncpt_incrs);
+    std::memcpy(syncPtIncrs.data(), input.data() + startPt,
+                sizeof(IoctlSyncPtIncr) * params.num_syncpt_incrs);
+    startPt += sizeof(IoctlSyncPtIncr) * params.num_syncpt_incrs;
+    // apply increment to sync points and create new one if not existed
+    for (auto syncIncr : syncPtIncrs) {
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
    for (const auto& syncIncr : syncPtIncrs) {
```

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.h [hunk 3]
```diff
@@ -23,16 +23,107 @@ class nvhost_nvdec final : public nvdevice {
...
+        // [in]
+        u32_le syncpt_id;
+        // [out] Returns the current syncpoint value for a given module. Identical to Linux driver.
+        u32_le syncpt_value;
+    };
+    static_assert(sizeof(IoctChannelSyncPoint) == 8, "IoctChannelSyncPoint is incorrect size");
+
     u32_le nvmap_fd{};
 
+    std::unordered_map<u32, u32> syncPtValues;
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.h]: `<unordered_map>` needs to be included.

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 4]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+    std::memcpy(handles.data(), input.data() + 12,
+                sizeof(IoctlHandleMapBuffer) * params.num_handles);
+
+    LOG_WARNING(Service_NVDRV, "(STUBBED) called, num_handles: {}, is_compressed: {}",
+                params.num_handles, params.is_compressed);
+
+    // TODO(namkazt): Uses nvmap_pin internally to pin a given number of nvmap handles to an
+    // appropriate device physical address.
+
+    std::memcpy(output.data(), &params, 12);
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
    std::memcpy(output.data(), &params, sizeof(params));
```

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 5]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+
+    params.waitbase_value = 0;
+
+    std::memcpy(output.data(), &params, output.size());
+    return 0;
+}
+
+u32 nvhost_nvdec::ChannelMapCmdBuffer(const std::vector<u8>& input, std::vector<u8>& output) {
+    IoctlMapCmdBuffer params{};
+    std::memcpy(&params, input.data(), 12);
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
    std::memcpy(&params, input.data(), sizeof(params));
```

**Comment by namkazt** [comment]
> Btw, yuzu implements sync points if you look at nv_host_ctrl you'll see how to read them.

thanks, i'm still not sure about it so will take a look.

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 6]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
     return 0;
 }
 
+u32 nvhost_nvdec::ChannelSubmit(const std::vector<u8>& input, std::vector<u8>& output) {
+    u32 startPt = 16;
+
+    IoctlSubmit params{};
+    std::memcpy(&params, input.data(), startPt);
+
+    std::vector<IoctlCmdBuf> cmdBufs(params.num_cmdbufs);
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
    std::vector<IoctlCmdBuf> cmd_bufs(params.num_cmdbufs);
```

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 7]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
     return 0;
 }
 
+u32 nvhost_nvdec::ChannelSubmit(const std::vector<u8>& input, std::vector<u8>& output) {
+    u32 startPt = 16;
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
    u32 start_point = 16;
```

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 8]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+    std::memcpy(cmdBufs.data(), input.data() + startPt, sizeof(IoctlCmdBuf) * params.num_cmdbufs);
+    startPt += sizeof(IoctlCmdBuf) * params.num_cmdbufs;
+    // TODO(namkazt): what to do with this?
+
+    std::vector<IoctlReloc> relocs(params.num_relocs);
+    std::memcpy(relocs.data(), input.data() + startPt, sizeof(IoctlReloc) * params.num_relocs);
+    startPt += sizeof(IoctlReloc) * params.num_relocs;
+    // TODO(namkazt): what to do with this?
+
+    std::vector<IoctlRelocShift> relocShifts(params.num_relocs);
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
    std::vector<IoctlRelocShift> reloc_shifts(params.num_relocs);
```

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 9]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+    std::memcpy(relocShifts.data(), input.data() + startPt,
+                sizeof(IoctlRelocShift) * params.num_relocs);
+    startPt += sizeof(IoctlRelocShift) * params.num_relocs;
+    // TODO(namkazt): what to do with this?
+
+    std::vector<IoctlSyncPtIncr> syncPtIncrs(params.num_syncpt_incrs);
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
    std::vector<IoctlSyncPtIncr> sync_pt_incrs(params.num_syncpt_incrs);
```

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 10]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+    std::memcpy(syncPtIncrs.data(), input.data() + startPt,
+                sizeof(IoctlSyncPtIncr) * params.num_syncpt_incrs);
+    startPt += sizeof(IoctlSyncPtIncr) * params.num_syncpt_incrs;
+    // apply increment to sync points and create new one if not existed
+    for (const auto& syncIncr : syncPtIncrs) {
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
    for (const auto& sync_incr : sync_pt_incrs) {
```

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 11]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+        auto itr = syncPtValues.find(syncIncr.syncpt_id);
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
        auto iter = syncPtValues.find(syncIncr.syncpt_id);
```

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 12]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+        if (itr == syncPtValues.end()) {
+            syncPtValues[syncIncr.syncpt_id] = syncIncr.syncpt_incrs;
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
            syncPtValues.insert_or_assign(syncIncr.syncpt_id, syncIncr.syncpt_incrs);
```

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 13]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+        } else {
+            syncPtValues[syncIncr.syncpt_id] += syncIncr.syncpt_incrs;
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]: You don't need to do a lookup into the map again. You can use the iterator here to increment the value

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp [hunk 14]
```diff
@@ -37,4 +54,96 @@ u32 nvhost_nvdec::SetNVMAPfd(const std::vector<u8>& input, std::vector<u8>& outp
...
+    std::memcpy(output.data(), &params, sizeof(params));
+    return 0;
+}
+
+u32 nvhost_nvdec::ChannelGetSyncPoint(const std::vector<u8>& input, std::vector<u8>& output) {
+    IoctChannelSyncPoint params{};
+    std::memcpy(&params, input.data(), input.size());
+    LOG_WARNING(Service_NVDRV, "called, module_id: {}", params.syncpt_id);
+
+    auto itr = syncPtValues.find(params.syncpt_id);
```
Comment (lioncash) [inline src/core/hle/service/nvdrv/devices/nvhost_nvdec.cpp]:
```suggestion
    const auto iter = syncPtValues.find(params.syncpt_id);
```

File: src/core/hle/service/nvdrv/devices/nvhost_nvdec.h (hunk 3 above)
Comment (lioncash) [reply src/core/hle/service/nvdrv/devices/nvhost_nvdec.h]: also `sync_pt_values`.

**Comment by ogniK5377** [comment]
You require the implementation of a command pusher for nvdec and vic. I have an implementation of it already working however there was various issues with h264 scaling and vp9 reference frames which still need to be looked into before nvdec should be considered. A few games also functioned quite odd with nvdec, one of those being links awakening becoming unplayable. More research needs to be done into it

**Comment by namkazt** [comment]
> You require the implementation of a command pusher for nvdec and vic. I have an implementation of it already working however there was various issues with h264 scaling and vp9 reference frames which still need to be looked into before nvdec should be considered. A few games also functioned quite odd with nvdec, one of those being links awakening becoming unplayable. More research needs to be done into it

Yes, but still is a part need to implement right?. I just trying to make it work so if you already implement it i really want to take a look :D

**Comment by ogniK5377** [comment]
> > You require the implementation of a command pusher for nvdec and vic. I have an implementation of it already working however there was various issues with h264 scaling and vp9 reference frames which still need to be looked into before nvdec should be considered. A few games also functioned quite odd with nvdec, one of those being links awakening becoming unplayable. More research needs to be done into it
> 
> Yes, but still is a part need to implement right?. I just trying to make it work so if you already implement it i really want to take a look :D @ogniK5377

It caused quite a few regressions and had too many issues, so the changes were never pushed. More reversing needs to be done before an implementation is pushed out

**Comment by namkazt** [comment]
Did you have any branch out there where it is pushed?.

**Comment by ogniK5377** [comment]
> Did you have any branch out there where it is pushed?.

Was on a private branch which was quite out of date, like I said it had many regressions in other games. We need more information to understand how it actually works before implementing it. Breaking too many games for video support isn't a good solution

**Comment by ogniK5377** [comment]
What's the status of this PR? Is this being researched or can it be closed?

**Comment by namkazt** [comment]
> What's the status of this PR? Is this being researched or can it be closed?

it can be closed

**Pull request closed without merging**